frontend/src/metabase-lib/Question.ts

```diff
@@ -660,20 +666,23 @@ class Question {
     const tableColumns = this.setting("table.columns");
     if (
       tableColumns &&
-      (addedColumnNames.length > 0 || removedColumnNames.length > 0)
+      (addedColumns.length > 0 || removedColumns.length > 0)
     ) {
       return this.updateSettings({
         "table.columns": [
           ...tableColumns.filter(
             column =>
-              !removedColumnNames.includes(column.name) &&
-              !addedColumnNames.includes(column.name),
+              !addedColumns.some(
+                ({ columnInfo }) => column.name === columnInfo.name,
```

Comment:
nit: maybe we should start mark datasetColumns as legacyColumns like with query?